### DISCUSSION TITLE
Shared and trusted data between middleware and server

### DISCUSSION THREAD

**Discussion by tinleym** [vercel/next.js#42879] [Ideas]
### Describe the feature you'd like to request

It would be nice if there was a way to share data between middleware and server and trust that the server-side read of those values can't be spoofed by a malicious client.

For example, it would be nice to get a userId from a cookie in middleware and then pass it to the server to preclude a second lookup.

### Describe the solution you'd like

1) Through @cache? Passing data through a request seems like spoof vector.

2) Or, via a request with needed firewalls downstream.

### Describe alternatives you've considered

1) Attaching the data on the request to be passed along. This seems insecure.

2) Refetching the data in each environment.

**Comment by milukukoz** [comment]
I also same case.
I'm using a singleton pattern to make sharing data between server components however it's not working correctly.
I hope someone can explain

init.ts(store data)
![image](https://user-images.githubusercontent.com/17758133/201579914-5120a2bf-5f1a-411c-82ed-dbafd4b04879.png)

middleware.ts(get cookie from client then make a request to merge data into store)
![image](https://user-images.githubusercontent.com/17758133/201580005-b0cbeb27-b578-4b3c-afbf-692e5fc17bba.png)

but this pattern does not keep the data
the logging when running
![image](https://user-images.githubusercontent.com/17758133/201580582-c92550f9-d5b5-4cd7-ac2e-e2119d63aa9d.png)
(the store has data that was merged from API)

but when loading the page
![image](https://user-images.githubusercontent.com/17758133/201580914-889b84f6-adf1-4810-9289-fcb751a45340.png)

**Reply by tinleym** [reply]
@milukukoz you need to attach data you're passing along to the server into the request field of NextResponse, which then needs to be returned. Here's an example: https://nextjs.org/docs/advanced-features/middleware#setting-headers
